### ISSUE TITLE
Implement Gradient Accumulation

### ISSUE THREAD

**Issue by waldheinz** [filipstrand/mflux#373]
I would like to implement gradient accumulation if this has a chance to get merged. I stumbled upon this:

https://github.com/ml-explore/mlx-examples/issues/929

And think it would not be too hard to add similar logic to the `TrainingTrainer.train` method. I imagine there would be a new block like

```py
if training_state.should_run_optimizer(training_spec):
    training_state.optimizer.optimizer.update(model=adapter.model(), gradients=accum_grads)
```

Inside the loop iterating the batches, with the training spec getting a new `grad_accum_steps` or whatever it will be called. Any thoughts on this?

**Comment by fxd0h** [comment]
implemented this for LoRA training on Apple Silicon: [a8e1cc6](https://github.com/fxd0h/mflux/commit/a8e1cc6) adds `gradient_accumulation_steps` to the optimizer config (averages grads over N micro-batches, steps once per window). it's part of [#447](https://github.com/filipstrand/mflux/pull/447). noticed in [#357](https://github.com/filipstrand/mflux/issues/357) you'd prototyped one too, so happy to compare notes or fold in anything yours does better.

**Comment by waldheinz** [comment]
Your commit a8e1cc6 is pretty much identical to what I have, one small thing: IIRC, you can just do

```
mx.eval(training_state.optimizer.optimizer.update(model=adapter.model(), gradients=grads))
```

instead of grabbing `training_state.optimizer.optimizer.state` (the update function returns the new state). This may be considered a little bit cleaner, but I can't verify right now.

**Comment by fxd0h** [comment]
thanks for taking a look. checked the eval tweak: `optimizer.update()` returns None in current mlx (not the new state), so `mx.eval(update(...))` would just eval None and the in-place update wouldn't get forced. the explicit `mx.eval(model.parameters(), optimizer.state)` is what pins it. so I kept that as-is, but good to confirm the rest matches yours.

**Comment by fxd0h** [comment]
Landed in #447, merged today. `gradient_accumulation_steps` lives in the training spec, and the trainer scales each micro-batch by the window, accumulates, and steps once the window closes.

One detail worth flagging because it is easy to get wrong: the window counts valid micro-batches rather than iterations, so a skipped non-finite batch extends it instead of closing it early on a half-magnitude gradient. Three tests on main cover a clean window, a skip, and a window that never completes.

Closing as done.
